### PR TITLE
Track `QuantumCircuit.global_phase` in `ParameterTable`

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

We have previously always had a split where circuit parameters used in instructions were tracked in the `ParameterTable`, but any parameters used in the global phase were not.  Any method that influenced the parameters needed to separately check the global phase, and merge that information with that in the `ParameterTable`.  This made it easy to forget, or easy for the handling of it to become out of sync.

This commit now tracks the global phase as part of the `ParameterTable`, so this object is now the canonical source of parameter information for the circuit (outside the context of calibrations, which are handled entirely separately).  The `ParameterTable` is an internal detail, and only accessible through private attributes, so is not part of the public interface.

### Details and comments

Dependent on #11109, which this PR currently includes.  I'll rebase it after that merges, but I'm putting this up a little early just to have things to link to in the dependency chain of #7107.

No changelog because it's a dev-only internal detail (in theory).

### CODE REVIEW

**Comment by jakelishman** [comment]
Now rebased over #11109.

**Review by ElePT — DISMISSED** [review]
From my (less limited than a few hours ago, but still limited) understanding, I don't see any functional issue with this implementation. It might not be the most straightforward, manual intervention is needed in a few points to make sure that the global phase is tracked properly (for example in `copy`), and this is adding responsibility to the developers (to keep on top of these details). But at the same time, I cannot come up with any concrete feedback or suggestions to improve it. I will approve to make sure this PR is not blocked, but it might be worth having somebody else give it some thought too.

**Comment by jakelishman** [comment]
Yeah, that's pretty much how I felt writing it as well, haha. I don't like what I've done hugely, but I don't have any better ideas, and overall this makes parameter-handling code more local.

File: qiskit/circuit/quantumcircuit.py
Comment (kevinhartman) [inline qiskit/circuit/quantumcircuit.py]: Is this no longer important to warn about?

**Review by kevinhartman — DISMISSED** [review]
LGTM.

I'll need to update #11214 once this merges to handle parameter remapping in `circuit_to_instruction`. 

Approving but holding off on merge in case addressing my comment warrants any code change. Feel free to merge as is.

File: qiskit/circuit/quantumcircuit.py
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py]: It was a mistake, thanks - when splitting up a larger commit that had originally gone further, I'd left this removed.  Fixed in 31009c9.

(Then I realised I made a mistake in removing it in the other commit anyway - it was a different `ParameterTable` method that stopped viewing the internals mutably, not the one here!)

**Review by kevinhartman — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
